### PR TITLE
Fix Gemma2 4d attention mask

### LINKED ISSUE #31673: Label Leakage in Gemma 2 Finetuning
### System Info

- `transformers` version: 4.42.1
- Platform: Linux-5.4.0-147-generic-x86_64-with-glibc2.31
- Python version: 3.10.13
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.30.1
- PyTorch version (GPU?): 2.3.0+cu121 (True)

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

#### Loss curves

We initially found that the training loss quickly dropped to 0 when fine-tuning the gemma2 models.

image

This phenomenon is very similar to label leakage.

#### Breakpoints

We changed the following lines of `Gemma2DecoderLayer` to the below form.

https://github.com/huggingface/transformers/blob/1c68f2cafb4ca54562f74b66d1085b68dd6682f5/src/transformers/models/gemma2/modeling_gemma2.py#L618-L633

```python
        if self.is_sliding and attention_mask is not None:  # efficient SDPA and no padding
            print("before", attention_mask)
            attention_mask = attention_mask * torch.tril(
                torch.ones_like(attention_mask), diagonal=-self.sliding_window
            )
            print("after", attention_mask)
            if attention_mask.shape[1] <= 1:  # when decoding
                attention_mask = attention_mask[:, -self.sliding_window :]
```

Then we execute the following code to trigger it.

```python
import torch
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2-9b-it", torch_dtype="auto", device_map="auto", attn_implementation="eager"
)
inputs = torch.tensor([[0, 1, 2]]).to(model.device)
labels = inputs.clone()
outputs = model(input_ids=inputs, labels=labels)
```

The result is:

```
before tensor([[[[ 0.0000e+00, -3.3895e+38, -3.3895e+38],
          [ 0.0000e+00,  0.0000e+00, -3.3895e+38],
          [ 0.0000e+00,  0.0000e+00,  0.0000e+00]]]], device='cuda:0',
       dtype=torch.bfloat16)
after tensor([[[[0., -0., -0.],
          [0., 0., -0.],
          [0., 0., 0.]]]], device='cuda:0', dtype=torch.bfloat16)
before tensor([[[[ 0.0000e+00, -3.3895e+38, -3.3895e+38],
          [ 0.0000e+00,  0.0000e+00, -3.3895e+38],
          [ 0.0000e+00,  0.0000e+00,  0.0000e+00]]]], device='cuda:0',
       dtype=torch.bfloat16)
after tensor([[[[0., -0., -0.],
          [0., 0., -0.],
          [0., 0., 0.]]]], device='cuda:0', dtype=torch.bfloat16)
...
```

### Expected behavior

The attention mask should not be all zeros; otherwise, it will use full attention instead of causal attention when calculating the loss, leading to label leakage.

### PR DESCRIPTION
**Pull request by hiyouga**
# What does this PR do?

(Might) continuation of #31670 

Fixes #31673 

The attention mask here should be a 4d mask, while the current implementation mistakenly treated it as a 2d mask. Additionally, the mask needs to be inverted to prevent label leakage.

Before this pr:

image

After this pr:

image2

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
Thanks, with this we do avoid the overflow!

**Comment by ArthurZucker** [comment]
The patch is out for this! Thanks again 🤗 https://pypi.org/project/transformers/4.42.2/

**Comment by kkk935208447** [comment]
Thank you all for your hard work. I've been dealing with this issue recently as well.
